Frameworks/Foundation/NSObject_NSKeyValueOrderedSetAdapter.mm

```diff
@@ -0,0 +1,353 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/Foundation.h>
+#import <Starboard.h>
+#import <Starboard/String.h>
+#import <Starboard/SmartTypes.h>
+
+#import "NSObject_NSKeyValueCoding-Internal.h"
+
+/*
+* KVC Ordered Set Adapters
+*
+* mutableOrderedSetValueForKey:@"xx" will ALWAYS return an NSOrderedSet-workalike. It'll do the following things:
+* - Preferentially use countOfXx and objectInXxAtIndex: for accessors.
+* - Preferentially use insertObject:inXxAtIndex: and -removeObjectFromXxAtIndex: for mutators.
+* - If these don't exist, accessors and mutators are proxied through to the following:
+*     - The getters getXx, xx, and isXx.
+*     - The instance Variables _xx or xx.
+*     - valueForKey:@"xx"
+* - If we used an ivar or valueForKey:, values of nil become empty NSOrderedSets.
+* - If we used a getter, a mutable copy of the object we get back is made.
+* - The mutation is attempted.
+*     - If we used an ivar, this will fail for immutable members.
+* - If a method mutated the copy of xx, setters cascade as follows:
+*     - The setter setXx:
+*     - Directly into the instance variable _xx or xx.
+*     - setValue: forKey:@"xx"
+*
+* Exceptions are expected at every stage of this pipeline.
+* if the ivar xx contains a NSNumber, we will attempt (fail) to get a mutable copy of it,
+* and then attempt (fail) to treat it like an array. This is entirely in keeping
+* with the behaviour of the reference platform.
+*
+* No caching is done, as, on the reference platform
+*
+* NSMutableOrderedSet* orderedSet = [object mutableOrderedSetValueForKey:@"xx"];
+* [object setValue:[NSSet orderedSetWithObject:@1024] forKey:@"xx"];
+*
+* [orderedSet containsObject:@1024];
+*
+* holds true.
+*/
+namespace {
+struct ProxyOrderedSet {
```

Comment:
This is all nearly the same as ProxyArray, but I kept them separate because I didn't think they should be sharing an implementation directly, though if we like I can minimize the amount of duplicated code and move these into a big shared implementation